src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -751,6 +815,12 @@ private Task DecodeTrailersAsync(bool endHeaders, Span<byte> payload)
 
         private void StartStream<TContext>(IHttpApplication<TContext> application)
         {
+            if (_stateManager.State == ConnectionState.Closed)
```

Comment:
StartStream and ResetRequestHeaderParsingState must be completed atomically

Reply:
Is it possible for this condition to be true for a well-behaved client? I'm guessing we've already sent a goaway frame at this point, so there isn't too much point in sending another. But if there's an unavoidable where even a well-behaved client can attempt to start a new stream while we're in the closed state, just not tracking the connection seems problematic. Any subsequent data frames sent by the client for that stream would cause the server to terminate the connection with an error immediately.

It also appears _streams could go from non-empty to empty back to non-empty all while the connection is the closing state. This would allow to connection to enter the closed state with a non-empty streams dictionary which seems problematic.

Reply:
The way this is written, it seem like theres a race where _streams could go from non-empty to empty back to non-empty all while the connection is the closing state. This would allow to connection to enter the closed state with a non-empty streams dictionary which seems problematic.

Reply:
This is possible. For example, the client is opening streams and sending new requests on them while the server is in the process of shutdown (via ctrl-c). Suppose the shutdown was completed relatively quickly on the server, there's a chance that by the time the client's request for new streams will arrive when the connection is already closed. This is a classic race condition described in the HTTP/2 spec. The mitigation is that when we close the connection, we need to send a final GOAWAY that has the highest frame number set. This way, the client that receives the GOAWAY will be able to tell which frames need to be retried.

As for the second situation, _streams should never be able to go from empty to non-empty. Once the connection reaches the Closing state and _streams become empty, the state will be changed to Closed. During the Closed state, no new streams can be started.

Reply:
> As for the second situation, _streams should never be able to go from empty to non-empty. Once the connection reaches the Closing state and _streams become empty, the state will be changed to Closed. During the Closed state, no new streams can be started.

I don't see how you can guarantee this without a lock.

Reply:
Never mind. I see this is now called with the lock acquired.

Reply:
Yea before I switched to the implementation using locks, it there was a race here.